File: ProblemSet/509.fibonacci-number/python/509.fibonacci-number 矩阵求幂法.py
#
# @lc app=leetcode id=509 lang=python3
#
# [509] Fibonacci Number
#

# @lc code=start
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        ans = [[1, 0], [0, 1]]
        a = [[1, 1], [1, 0]]
        n -= 1
        while n > 0:
            if n & 1:
                self.matrix_mul(ans, a)
            self.matrix_mul(a, a)
            n >>= 1
        return ans[0][0]
        
    def matrix_mul(self, a: List[List[int]], b: List[List[int]]):
        x11 = a[0][0] * b[0][0] + a[0][1] * b[1][0]
        x12 = a[0][0] * b[0][1] + a[0][1] * b[1][1]
        x21 = a[1][0] * b[0][0] + a[1][1] * b[1][0]
        x22 = a[1][0] * b[0][1] + a[1][1] * b[1][1]

        a[0][0], a[0][1], a[1][0], a[1][1] = x11, x12, x21, x22
        
# @lc code=end

File: ProblemSet/509.fibonacci-number/java/509.fibonacci-number 矩阵求幂法.java
/*
 * @lc app=leetcode id=509 lang=java
 *
 * [509] Fibonacci Number
 */

// @lc code=start
class Solution {
    public int fib(int N) {
        if (N <= 1) return N;
        int[][] ans = new int[][]{{1, 0}, {0, 1}};
        int[][] A = new int[][]{{1, 1}, {1, 0}};
        int n = N - 1;
        while (n > 0) {
            if ((n & 1) == 1) matrixMul(ans, A);
            matrixMul(A, A);
            n >>= 1;
        }
        return ans[0][0];
    }

    private void matrixMul(int[][] A, int[][] B) {
        int x11 = A[0][0] * B[0][0] + A[0][1] * B[1][0];
        int x12 = A[0][0] * B[0][1] + A[0][1] * B[1][1];
        int x21 = A[1][0] * B[0][0] + A[1][1] * B[1][0];
        int x22 = A[1][0] * B[0][1] + A[1][1] * B[1][1];

        A[0][0] = x11;
        A[0][1] = x12;
        A[1][0] = x21;
        A[1][1] = x22;
    }
}
// @lc code=end

